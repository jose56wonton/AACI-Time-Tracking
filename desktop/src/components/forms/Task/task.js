import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { Field, Form } from 'formik';
import cx from 'classnames';
import { withStyles } from '@material-ui/core/styles';
import { Grid,Tooltip,IconButton, Typography, Button } from '@material-ui/core';
import { Delete } from '@material-ui/icons';

import styles from './styles';
import TextField from 'components/inputs/TextField';
import Select from 'components/inputs/Select';
import Switch from 'components/inputs/Switch';

export class Task extends Component {
  
  render() {
    //console.log(this.props);
    const { label,removeTask,type,classes,categories, subcategories,isSubmitting,resetForm,initialValues,errors } = this.props;
    return (
      <Form>
        <Grid container spacing={24} className={classes.gridContainer}>
          <Grid item xs={12}  className={cx(classes.headerRow,classes.row)} >
            <Typography variant="h6">{label}</Typography>
            {type === `edit` && (
              <Tooltip title="Delete">
                <IconButton onClick={removeTask} aria-label="Delete">
                  <Delete />
                </IconButton>
              </Tooltip>
            )}
          </Grid>
          <Grid item xs={12}  className={classes.row}>
            <Field
              name="name"
              component={TextField}
              margin="none"
              label="Task Name"
              type="search"
              className={classes.field}
              helper="normal"
            />
            <Field
              name="isActive"
              component={Switch}
              label="Active"
              className={classes.field}
            />
          </Grid>
          <Grid item xs={12} className={classes.row}>
            <Field
              name="categoryId"
              component={Select}
              items={categories}
              fullWidth
              label="Category"
              className={classes.field}
            />
            <Field
              name="subcategoryId"
              component={Select}
              items={subcategories.filter(subcat => {
                return subcat.categoryId === this.props.values.categoryId;
              })}
              fullWidth
              label="Subcategory"
              className={classes.field}
            />            
          </Grid>          
          <Grid item xs={12} className={classes.row}>
            <Typography
              color="error"
              variant="button"
              className={classes.field}
            >
              {errors.submit}
            </Typography>
            <div>
              <Button
                type="submit"
                color="primary"
                disabled={isSubmitting || Object.keys(errors).length !== 0}
                variant="contained"
                className={classes.button}
              >
                Save
              </Button>
              <Button
                onClick={() => {
                  resetForm(initialValues);
                }}
                id="task-reset-button"
                disabled={isSubmitting}
                color="secondary"
                variant="text"
                className={classes.button}
              >
                Reset
              </Button>
            </div>
          </Grid>        
        </Grid>
      </Form>
    );
  }
}

Task.propTypes = {
  classes: PropTypes.object.isRequired,
  label: PropTypes.string.isRequired,
  removeTask: PropTypes.func,
  type: PropTypes.string.isRequired,
  categories: PropTypes.array.isRequired,
  subcategories: PropTypes.array.isRequired,
  isSubmitting: PropTypes.bool.isRequired,
  resetForm: PropTypes.func.isRequired,
  initialValues: PropTypes.object.isRequired,
  errors: PropTypes.object.isRequired,
  values: PropTypes.object
};

export default withStyles(styles)(Task);