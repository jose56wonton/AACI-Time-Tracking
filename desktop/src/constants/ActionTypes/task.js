export const GET_TASKS_REQUEST = 'get_tasks_request';
export const GET_TASKS_SUCCESS = 'get_tasks_success';
export const GET_TASKS_FAILURE = 'get_tasks_failure';

export const DELETE_TASK_REQUEST = 'delete_task_request';
export const DELETE_TASK_SUCCESS = 'delete_task_success';
export const DELETE_TASK_FAILURE = 'delete_task_failure';

export const CREATE_TASK_REQUEST = 'create_task_request';
export const CREATE_TASK_SUCCESS = 'create_task_success';
export const CREATE_TASK_FAILURE = 'create_task_failure';

export const UPDATE_TASK_REQUEST = 'update_task_request';
export const UPDATE_TASK_SUCCESS = 'update_task_success';
export const UPDATE_TASK_FAILURE = 'update_task_failure';