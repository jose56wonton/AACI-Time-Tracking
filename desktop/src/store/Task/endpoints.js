//import { HOST } from 'constants/network';
import { generateCRUDEndpoints } from 'helpers/endpoint.helper';
//import axios from 'axios';
const DOMAIN = `tasks`;

// const get = () => {
//   return axios.get(`${HOST}/${DOMAIN}?filter[include][subcategory]=dimension&filter[include][subcategory]=category`);
// };
const CRUDEndpoints = generateCRUDEndpoints(DOMAIN);

export default {
  ...CRUDEndpoints
};