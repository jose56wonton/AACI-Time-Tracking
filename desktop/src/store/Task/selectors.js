import { createSelector } from 'reselect';

export const getTasksFromEntities = state => state.entities.tasks;
export const getTasksFromResults = state => state.results.tasks;

export const getDimensionsFromEntities = state => state.entities.dimensions;
export const getCategoriesFromEntities = state => state.entities.categories;
export const getSubcategoriesFromEntities = state => state.entities.subcategories;

export const getAnalyzeState = state => state.analyze;

export const getAllTasks = createSelector(
  getTasksFromEntities,
  getTasksFromResults,
  (tasks, results) => {
    if (!results || results.size === 0) return null;
    return results.map(taskId => {
      return tasks[taskId];
    });
  },
);



export const getAllTasksWithContent = createSelector(
  getTasksFromEntities,
  getTasksFromResults,
  getDimensionsFromEntities,
  getCategoriesFromEntities,
  getSubcategoriesFromEntities,
  (tasks,results,dimensions,categories,subcategories) => {
    if(!results || results.size === 0) return null;
    //console.log(dimensions, categories);
    return results.map(taskId => {
      const task = tasks[taskId];
      return {
        ...task,
        subcategory: subcategories[task.subcategoryId],
        category: categories[subcategories[task.subcategoryId].categoryId],
        dimension: dimensions[subcategories[task.subcategoryId].dimensionId] 
      };
    });
  }
);

export const getAllTaskObjects = createSelector(
  getAllTasksWithContent,
  tasks => {
    // if the task array is empty
    if (!tasks) return null;    
    // reduce the task array to a object with id as they key
    return Object.assign({}, ...tasks.map(object => ({ [object.id]: object })));    
  },
);

export const getSelectedTask = createSelector(
  getTasksFromEntities,
  getSubcategoriesFromEntities,
  getAnalyzeState,
  (tasks,subcategories,analyze) => {
    if(analyze.task === -1)
      return {};
    else{
      const temp = tasks[analyze.task];
      return { ...temp,categoryId: subcategories[temp.subcategoryId].categoryId };
    }      
  }
);