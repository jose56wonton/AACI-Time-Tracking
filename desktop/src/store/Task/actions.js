import { taskActionTypes } from 'constants/ActionTypes';

import * as endpoint from './endpoints';
import { normalize } from 'normalizr';
import * as schemas from 'store/schemas';
import { snackActions } from 'store/actions';
import * as status from 'constants/status';
import { normalizeEmbeddedData } from 'helpers/store';

export const getTasks = () => {
  return async dispatch => {
    dispatch({ type: taskActionTypes.GET_TASKS_REQUEST });
    try {
  
      const response = await endpoint.getTasks(); 
      let payload = normalize({ tasks: response.data }, schemas.taskArray);

      return dispatch({
        type: taskActionTypes.GET_TASKS_SUCCESS,
        payload: normalizeEmbeddedData(payload)
      });
    } catch (e) {
      dispatch({
        type: taskActionTypes.GET_TASKS_FAILURE,
        payload: e
      });
      throw e;
    }
  };
};

export const postTask = task => {
  return async dispatch => {
    dispatch({ type: taskActionTypes.CREATE_TASK_REQUEST });
    try {
      const response = await endpoint.postTask(task);
      const payload = normalize({ tasks: [response.data] }, schemas.taskArray);

      await dispatch(
        snackActions.openSnack(status.SUCCESS, 'Task Created'),
      );

      return dispatch({
        type: taskActionTypes.CREATE_TASK_SUCCESS,
        payload
      });
    } catch (e) {
      console.log(e);
      return dispatch({
        type: taskActionTypes.CREATE_TASK_FAILURE,
        payload: e
      });
    }
  };
};
export const putTask = task => {
  return async dispatch => {
    dispatch({ type: taskActionTypes.UPDATE_TASK_REQUEST });
    try {
      const response = await endpoint.putTask(task.id,task);
      const payload = normalize({ tasks: [response.data] }, schemas.taskArray);

      await dispatch(
        snackActions.openSnack(status.SUCCESS, 'Task Updated'),
      );

      return dispatch({
        type: taskActionTypes.UPDATE_TASK_SUCCESS,
        payload
      });
    } catch (e) {
      console.log(e);
      return dispatch({
        type: taskActionTypes.UPDATE_TASK_FAILURE,
        payload: e
      });
    }
  };
};
export const deleteTask = task => {
  return async dispatch => {
    dispatch({ type: taskActionTypes.DELETE_TASK_REQUEST });
    try {
      const response = await endpoint.deleteTask(task);
      const payload = normalize({ tasks: [response.data] }, schemas.taskArray);

      await dispatch(
        snackActions.openSnack(status.SUCCESS, 'Task deleted'),
      );

      const deleted = {
        entities:{
          tasks: [task.id]
        },
        result: {
          tasks: [task.id]
        }
      }

      return dispatch({
        type: taskActionTypes.DELETE_TASK_SUCCESS,
        deleted
      });
    } catch (e) {
      console.log(e);
      await dispatch(
        snackActions.openSnack(status.FAILURE, 'Task deletion failed!'),
      );
      return dispatch({
        type: taskActionTypes.DELETE_TASK_FAILURE,
        payload: e
      });
    }
  };
};