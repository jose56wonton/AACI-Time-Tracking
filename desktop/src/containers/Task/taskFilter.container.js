import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { Formik } from 'formik';
import { Card } from '@material-ui/core';

import Task from 'components/forms/Task';

import { categorySelectors,subcategorySelectors } from 'store/selectors';
import { analyzeActions } from 'store/actions';

import domain from 'constants/domains';

export class TaskDetail extends Component {  
  render() {
    const { categories, subcategories, clearFilter, taskFilters, taskFilterVisible } = this.props;
    
    if(taskFilterVisible){
      return (
        <Formik 
          enableReinitialize
          initialValues={{
            ...taskFilters
          }}          
          onSubmit={(values,formikFunctions) => {
            this.props.updateFilter({ // TODO: this should close the filter
              ...values            
            });
            formikFunctions.resetForm();  
          }}
          render={formikProps => {
            return (
              <Card style={{ position: `absolute`, top: `70px`, left: `2.5%`, zIndex: 900, width: `95%`, minHeight: `100px` }}>                
                <Task 
                  removeTask={this.removeTask}
                  categories={[{ id: -1, type: `All` }, ...categories]}
                  subcategories={[{ id: -1, type: `All`,categoryId: -1 }, ...subcategories]}                
                  label="Filter"
                  type="filter"
                  clearFilter={clearFilter}
                  {...formikProps}
                />
              </Card>
            );
          }}
        />
      );   
    }
    return null;
  }
}

/* istanbul ignore next */
const mapStateToProps = state => {
  return {
    taskFilters: state.analyze.taskFilters,
    taskFilterVisible: state.analyze.taskFilterVisible,
    status: state.analyze.taskStatus,
    categories: categorySelectors.getAllCategories(state),
    subcategories: subcategorySelectors.getAllSubcategories(state)    
  };
};

/* istanbul ignore next */
const mapDispatchToProps = dispatch => {
  return {
    updateFilter: filters => dispatch(analyzeActions.updateFilter(domain.TASK,filters)),
    clearFilter: () => dispatch(analyzeActions.clearFilter(domain.TASK))
  };
};

TaskDetail.propTypes = {
  status: PropTypes.string.isRequired
};

export default connect(mapStateToProps,mapDispatchToProps)(TaskDetail);