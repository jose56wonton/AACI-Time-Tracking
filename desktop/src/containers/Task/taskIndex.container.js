import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

// TODO: double check that these imports are ordered well
import { analyzeActions } from 'store/actions';
import { taskSelectors } from 'store/selectors';
import VirtualizedSortSelect from 'components/tables/VirtualizedSortSelect';
import Progress from 'components/helpers/Progress';
import * as TableDataTypes from 'constants/tableDataTypes';
import { analyzeStatus } from 'constants/analyze';
import domain from 'constants/domains';

export class TaskIndex extends Component {  
  selectLabel = selected =>`${selected.name} selected`;

  select = object => this.props.select(domain.TASK,object)

  add = () => this.props.setStatus(domain.TASK,analyzeStatus.ADDING)

  render() {
    const { tasks, selected } = this.props;

    if (!tasks) return <Progress variant="circular" fullWidth fullHeight />;
    
    return (
     
      <VirtualizedSortSelect 
        data={tasks || []}
        columns={rows}
        selected={selected}
        select={this.select}       
        initialSortBy='name'
      />
    );
  }
}

TaskIndex.propTypes = {
  tasks: PropTypes.array,
  selected: PropTypes.object,
  select: PropTypes.func.isRequired,
  setStatus: PropTypes.func.isRequired
};

/* istanbul ignore next */
const mapStateToProps = state => {
  return {
    tasks: taskSelectors.getAllTasksWithContent(state),
    selected: taskSelectors.getSelectedTask(state)
  };
};

/* istanbul ignore next */
const mapDispatchToProps = dispatch => {
  return {
    ...bindActionCreators({ ...analyzeActions }, dispatch)   
  };
};

export default connect( mapStateToProps, mapDispatchToProps)(TaskIndex);

const rows = [
  {
    id: `name`, 
    dataKey: `name`, 
    width: 150, 
    height: 56,     
    padding: `dense`,
    label: `Name`,
    type: TableDataTypes.STRING
  }, 
  {
    id: `category_type`,  
    dataKey: `category`, 
    width: 150, 
    height: 56,    
    padding: `dense`,
    label: `Category`,
    type: TableDataTypes.OBJECT,
    keys: [`type`]
  },
  {
    id: `subcategory_type`,  
    dataKey: `subcategory`, 
    width: 150, 
    height: 56,    
    padding: `dense`,
    label: `Subcategory`,
    type: TableDataTypes.OBJECT,
    keys: [`type`]
  },
 
  // { // ICEBOX: add dimension functionality
  //   id: `dimension_type`,   
  //   dataKey: `dimension`, 
  //   width: 150, 
  //   height: 56,  
  //   padding: `dense`,
  //   label: `Dimension`,
  //   type: TableDataTypes.OBJECT,
  //   keys: [`type`]
  // }, 
  {
    id: `isActive`,  
    dataKey: `isActive`,  // TODO: does align right look ok?
    width: 150, 
    height: 56,   
    align: `right`, 
    padding: `dense`,
    label: `Active`,
    type: TableDataTypes.BOOLEAN
  }
];
