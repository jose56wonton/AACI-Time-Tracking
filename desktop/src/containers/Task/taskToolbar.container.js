import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { analyzeActions } from 'store/actions';
import { taskSelectors } from 'store/selectors';
import SortSelectTable from 'components/tables/SortSelect';
import Progress from 'components/helpers/Progress';
import * as TableDataTypes from 'constants/tableDataTypes';
import { analyzeStatus } from 'constants/analyze';
import domain from 'constants/domains';

import AnalyzeToolbar from 'components/tables/toolbars/NewAnalyzeToolbar';

export class TaskToolbar extends Component {  
  selectLabel = selected =>`${selected.name} selected`;

  select = object => this.props.select(domain.TASK,object)
  // TODO: double check that all these methods are needed
  add = () => {
    const { selected, select, setStatus } = this.props;
    if(selected && selected.id){
      select(domain.TASK,selected.id);
    }
    setStatus(domain.TASK,analyzeStatus.ADDING)
  }
  render() {
    const { tasks, selected, toggleTaskFilter, taskFilterVisible } = this.props;
    // console.log('task analyze render');
    if (!tasks) return <Progress variant="circular" fullWidth fullHeight />;
    
    return (
      <AnalyzeToolbar 
        selectLabel={this.selectLabel}
        label="Task"
        add={this.add}
        selected={selected}
        toggleFilter={toggleTaskFilter}
        isFilterVisible={taskFilterVisible}
      />
    );
  }
}

TaskToolbar.propTypes = {
  tasks: PropTypes.array,
  selected: PropTypes.object,
  select: PropTypes.func.isRequired,
  setStatus: PropTypes.func.isRequired,
  toggleTaskFilter: PropTypes.func,
  taskFilterVisible: PropTypes.bool
};

/* istanbul ignore next */
const mapStateToProps = state => {
  return {
    taskFilterVisible: state.analyze.taskFilterVisible,
    tasks: taskSelectors.getAllTasksWithContent(state),
    selected: taskSelectors.getSelectedTask(state)
  };
};

/* istanbul ignore next */
const mapDispatchToProps = dispatch => {
  return {
    ...bindActionCreators({ ...analyzeActions }, dispatch),  
    toggleTaskFilter: () => dispatch(analyzeActions.toggleFilter(domain.TASK)) 
  };
};

export default connect( mapStateToProps, mapDispatchToProps)(TaskToolbar);
