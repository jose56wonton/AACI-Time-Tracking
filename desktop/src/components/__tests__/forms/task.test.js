import React from 'react';
import { shallow } from 'enzyme';

import { Task } from 'components/forms/Task/task';
import TaskHOC from 'components/forms/Task';

const props =  {  
  classes: {},
  label: 'label',
  removeTask: jest.fn(),
  type: 'edit',
  categories: [{ id: 0 },{ id: 1 },{ id: 2 }],
  subcategories: [{ id: 0 },{ id: 1 },{ id: 2 }],
  resetForm: jest.fn(),
  initialValues: {},
  isSubmitting: true,
  values: {
    pin: '123'
  },
  editCategories: jest.fn(),
  errors: {}
};

const setup = overRides => {  
  return shallow(<Task {...props} {...overRides}/>);    
};

const setupHOC = overRides => {
  return shallow(<TaskHOC {...props} {...overRides}/>);
};

describe('Task Component', () => {
  it('should render correctly', () => {
    setup();
  });
  it('should render correctly withStyles', () => {
    setupHOC();
  });
  it('should resetForm when #task-reset-button is clicked', () => {
    const wrapper = setup();
    const instance = wrapper.instance();
    instance.resetForm = jest.fn();
    expect(props.resetForm).toHaveBeenCalledTimes(0);
    wrapper.find('#task-reset-button').first().simulate('click');
    expect(props.resetForm).toHaveBeenCalledTimes(1);
  });  
});