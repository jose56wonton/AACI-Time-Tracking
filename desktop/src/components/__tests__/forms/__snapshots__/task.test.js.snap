// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Component should render correctly 1`] = `
<Form>
  <WithStyles(Grid)
    container={true}
    spacing={24}
  >
    <WithStyles(Grid)
      className=""
      item={true}
      xs={12}
    >
      <WithStyles(Typography)
        variant="h6"
      >
        label
      </WithStyles(Typography)>
      <WithStyles(Tooltip)
        title="Delete"
      >
        <WithStyles(IconButton)
          aria-label="Delete"
          onClick={[Function]}
        >
          <pure(DeleteIcon) />
        </WithStyles(IconButton)>
      </WithStyles(Tooltip)>
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      xs={12}
    >
      <C
        component={[Function]}
        helper="normal"
        label="Task Name"
        margin="none"
        name="name"
        type="search"
      />
      <C
        component={[Function]}
        label="Active"
        name="isActive"
      />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      xs={12}
    >
      <C
        component={[Function]}
        fullWidth={true}
        items={
          Array [
            Object {
              "id": 0,
            },
            Object {
              "id": 1,
            },
            Object {
              "id": 2,
            },
          ]
        }
        label="Category"
        name="categoryId"
      />
      <C
        component={[Function]}
        fullWidth={true}
        items={
          Array [
            Object {
              "id": 0,
            },
            Object {
              "id": 1,
            },
            Object {
              "id": 2,
            },
          ]
        }
        label="Subcategory"
        name="subcategoryId"
      />
      <WithStyles(Tooltip)
        title="Edit Categories"
      >
        <WithStyles(IconButton)
          aria-label="Edit Categories"
          onClick={[Function]}
        >
          <pure(EditIcon) />
        </WithStyles(IconButton)>
      </WithStyles(Tooltip)>
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      xs={12}
    >
      <WithStyles(Typography)
        color="error"
        variant="button"
      />
      <div>
        <WithStyles(Button)
          color="primary"
          disabled={true}
          type="submit"
          variant="contained"
        >
          Save
        </WithStyles(Button)>
        <WithStyles(Button)
          color="secondary"
          disabled={true}
          id="task-reset-button"
          onClick={[Function]}
          variant="text"
        >
          Reset
        </WithStyles(Button)>
      </div>
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</Form>
`;

exports[`Task Component should render correctly withStyles 1`] = `
<Task
  categories={
    Array [
      Object {
        "id": 0,
      },
      Object {
        "id": 1,
      },
      Object {
        "id": 2,
      },
    ]
  }
  classes={
    Object {
      "button": "Task-button-7",
      "error": "Task-error-6",
      "field": "Task-field-4",
      "gridContainer": "Task-gridContainer-1",
      "headerRow": "Task-headerRow-3",
      "row": "Task-row-2",
      "switchBox": "Task-switchBox-5",
    }
  }
  editCategories={[Function]}
  errors={Object {}}
  initialValues={Object {}}
  isSubmitting={true}
  label="label"
  removeTask={[Function]}
  resetForm={[Function]}
  subcategories={
    Array [
      Object {
        "id": 0,
      },
      Object {
        "id": 1,
      },
      Object {
        "id": 2,
      },
    ]
  }
  type="edit"
  values={
    Object {
      "pin": "123",
    }
  }
/>
`;
